android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java

```diff
@@ -26,76 +30,257 @@
+    private static final int UPLOAD_EXPIRY_WINDOW_MINUTES = 30;
+
     private final ForConsentedUsersApi api;
     private final StudyUploadEncryptor encryptor;
-    private final OkHttpClient s3OkhttpClient;
+    private final UploadDAO uploadDAO;
+    private final OkHttpClient okHttpClient = new OkHttpClient.Builder()
+            .connectTimeout(30, TimeUnit.SECONDS)
+            .readTimeout(30, TimeUnit.SECONDS)
+            .writeTimeout(30, TimeUnit.SECONDS)
+            .retryOnConnectionFailure(false).build();
 
     public UploadManager(AuthenticationManager authenticationManager, StudyUploadEncryptor
-            encryptor, OkHttpClient s3OkhttpClient) {
+            encryptor, UploadDAO uploadDAO) {
         this.api = authenticationManager.getApi();
         this.encryptor = encryptor;
-        this.s3OkhttpClient = s3OkhttpClient;
+        this.uploadDAO = uploadDAO;
     }
 
-    public Single<UploadValidationStatus> getStatus(String uploadId) {
-        return RxUtils.toBodySingle(api.getUploadStatus(uploadId));
+    public static class UploadFile {
+        public String filename;
+        public String contentType;
+        public long fileLength;
+        public String md5Hash;
+        public DateTime createdOn;
     }
 
+    /**
+     * Persists the archive on disk and add it to the queue of pending uploads, runs in an IO thread.
+     *
+     * @param filename filename for the archive
+     * @param archive  archive to be queued
+     * @return information about file produced from the archive
+     */
     @NonNull
-    public Single<UploadValidationStatus> upload(String filename, Archive archive) {
-        Single<UploadFile> uploadFileSingle = Single
-                .fromCallable(() -> persist(filename, archive)).cache();
+    public Single<UploadFile> queueUpload(String filename, Archive archive) {
+        return Single.fromCallable(() -> persist(filename, archive))
+                .subscribeOn(Schedulers.io());
+    }
 
-        Single<UploadSession> uploadSessionSingle = uploadFileSingle
-                .flatMap(this::requestUploadSession);
+    /**
+     * This triggers an attempt to upload all the previously queued upload files
+     * You must call queueUpload before this method will have anything to operate on
+     *
+     * @return Observable with information on if the upload was successful or not
+     */
+    @NonNull
+    public Completable processUploadFiles() {
+        Observable<String> filenameObservable = Observable.from(uploadDAO
+                .listUploadFilenames());
+
+        return filenameObservable
+                .map(filename -> uploadDAO.getUploadFile(filename))
+                .flatMap(uploadFile -> processUploadFile(uploadFile).toObservable())
+                .toCompletable();
+    }
 
-        return Single.zip(uploadFileSingle, uploadSessionSingle, this::uploadToS3).flatMap(i -> i);
+    /**
+     * Performs next upload step for a file.
```

Comment:
Honestly, this doesn't really help me understand what is going on in the method.